File: queue/smashRocks.py
class Solution:
    def lastStoneWeight(self, stones: List[int]) -> int:
        #heapq.heapify(stones)
        stones.sort() # O(nlog(n))
        i = len(stones) - 1
        while (i >= 1):
            # y >= x
            y = stones[i]
            x = stones[i-1]
            if (x == y):
                stones.pop()
                stones.pop()
                i -= 2
            else: # y > x
                newStone = y - x
                stones.pop()
                stones.pop()
                stones.append(newStone)
                # this part could be optimized using heap structure
                stones.sort() # --> O( < n * nlog(n))
                i -= 1
        if (len(stones) < 1):
            return 0
        return stones[0]
        

File: queue/smashRocks.js
/**
 * @param {number[]} stones
 * @return {number}
 */
var lastStoneWeight = function(stones) {
    stones.sort((x, y) => {
        return x-y;
    });
    let i = stones.length - 1;
    while (i >= 1) {
        // y >= x
        y = stones[i];
        x = stones[i-1];
        if (x===y) {
            stones.pop();
            stones.pop();
            i -= 2;
        }
        else { // y > x
            stones.pop();
            stones[stones.length - 1] = y - x;
            stones.sort((x, y) => {
                return x-y;
            });
            i -= 1;
        }
    }
    if (stones.length < 1) {
        return 0;
    }
    return stones[0];
};

